### PR TITLE
Build lz4 with meson to support VS2022

### PR DESCRIPTION
**Pull request by danyeaw**

### CODE REVIEW

**Comment by nacho** [comment]
hmmm is the meson build system official?

**Comment by pbor** [comment]
It is in the official repo, but it did not seem to get any recent commits... I think we would need to double check it is functionally equivalent, and in particular that it does not introduce performance regressions (eg in something as tweaked as a compression lib, missing compiler flags or not detecting an assembler dep could be problematic)

File: gvsbuild/projects/lz4.py [hunk 1]
```diff
@@ -15,31 +15,25 @@
...
         Project.__init__(
             self,
             "lz4",
             archive_url="https://github.com/lz4/lz4/archive/v1.9.3.tar.gz",
             archive_file_name="lz4-1.9.3.tar.gz",
             hash="030644df4611007ff7dc962d981f390361e6c97a34e5cbc393ddfbe019ffe2c1",
+            dependencies=["ninja", "meson"],
         )
+        self.add_param("-Ddefault_library=shared")
+        self.add_param("-Dbin_programs=true")
```
Comment (tristan957) [inline gvsbuild/projects/lz4.py:35]: Double check your options here

**Comment by tristan957** [comment]
Hi, I am the most recent committer of the Meson build in lz4. The Meson build is community-maintained for the most part, hence the upstream `contrib/` directory. I can confirm that this build works however and has the nod of Meson maintainer @eli-schwartz. We are currently waiting for an lz4 release to get the Meson build usable in the WrapDB.

**Comment by eli-schwartz** [comment]
Upstream officially only supports Makefile, this is the same policy that the upstream maintainer uses in e.g. zstd as well. All non-Makefile builds are considered equally unofficial and left to the community to maintain.

File: gvsbuild/projects/lz4.py (hunk 1 above)
Comment (danyeaw) [reply gvsbuild/projects/lz4.py:35]: Thanks. I was just copying the build options in the README. I submitted a PR to correct it: https://github.com/lz4/lz4/pull/1075

**Comment by pbor** [comment]
Thanks for the clarifications... I am ok using meson since it as unofficial as the manual VS projects and definitely nicer for maintenance.

I would still like to double check that we are not missing any performance-impacting compiler options.

With regard to `self.add_param("-Dprograms=true")`, maybe we should actually turn off binaries... I believe we are only interested in the library, and given the difference license, I would build only the lib by default

**Comment by danyeaw** [comment]
> With regard to `self.add_param("-Dprograms=true")`, maybe we should actually turn off binaries... I believe we are only interested in the library, and given the difference license, I would build only the lib by default

Good idea, I removed this parameter.

I am getting an error installing files, it looks like it is compiling lz4-1.dll, but trying to copy lz4.dll. Any ideas anyone?

```
PS C:\gtk-build\github\gvsbuild> python .\build.py -v build -p x64 --clean --msys-dir C:\tools\msys64 lz4 --vs-install-path "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools"
Building project lz4 (1.9.3)
Generating meson directory
The Meson build system
Version: 0.59.4
Source dir: C:\gtk-build\build\x64\release\lz4
Build dir: C:\gtk-build\build\x64\release\lz4\_gvsbuild-meson
Build type: native build
Project name: lz4
Project version: DUMMY
C compiler for the host machine: cl (msvc 19.31.31105 "Microsoft (R) C/C++ Optimizing Compiler Version 19.31.31105 for x64")
C linker for the host machine: link link 14.31.31105.0
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program GetLz4LibraryVersion.py found: YES (C:\gtk-build\tools\python.3.10.4\tools\python.exe C:\gtk-build\build\x64\release\lz4\meson\GetLz4LibraryVersion.py)
Message: Project version is now: 1.9.3
Compiler for C supports arguments -Wcast-qual: NO 
Compiler for C supports arguments -Wcast-align: NO 
Compiler for C supports arguments -Wshadow: NO 
Compiler for C supports arguments -Wswitch-enum: NO 
Compiler for C supports arguments -Wdeclaration-after-statement: NO 
Compiler for C supports arguments -Wstrict-prototypes: NO 
Compiler for C supports arguments -Wundef: NO 
Compiler for C supports arguments -Wpointer-arith: NO 
Compiler for C supports arguments -Wstrict-aliasing=1: NO 
Compiler for C supports arguments -DLZ4_DEBUG=1: YES 
Build targets in project: 1

Option buildtype is: debugoptimized [default: release]
Found ninja-1.8.2 at C:\gtk-build\tools\ninja-1.8.2\ninja.EXE
[5/6] Installing files.
Installing meson\lib\lz4-1.dll to C:\gtk-build\gtk\x64\release\bin
Traceback (most recent call last):
  File "C:\gtk-build\tools\meson-0.59.4\mesonbuild\mesonmain.py", line 228, in run
    return options.run_func(options)
  File "C:\gtk-build\tools\meson-0.59.4\mesonbuild\minstall.py", line 720, in run
    installer.do_install(datafilename)
  File "C:\gtk-build\tools\meson-0.59.4\mesonbuild\minstall.py", line 512, in do_install
    self.install_targets(d, dm, destdir, fullprefix)
  File "C:\gtk-build\tools\meson-0.59.4\mesonbuild\minstall.py", line 615, in install_targets
    raise RuntimeError(f'File {t.fname!r} could not be found')
RuntimeError: File 'meson\\lib\\lz4.lib' could not be found
FAILED: meson-install 
"C:\gtk-build\tools\python.3.10.4\tools\python.exe" "C:\gtk-build\tools\meson-0.59.4\meson.py" "install" "--no-rebuild"
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "C:\gtk-build\github\gvsbuild\gvsbuild\utils\builder.py", line 552, in build
    if self.__build_one(p):
  File "C:\gtk-build\github\gvsbuild\gvsbuild\utils\builder.py", line 688, in __build_one
    skip_deps = proj.build()
  File "C:\gtk-build\github\gvsbuild\gvsbuild\projects\lz4.py", line 37, in build
    Meson.build(self)
  File "C:\gtk-build\github\gvsbuild\gvsbuild\utils\base_builders.py", line 87, in build
    self.builder.exec_ninja(params="install", working_dir=ninja_build)
  File "C:\gtk-build\github\gvsbuild\gvsbuild\utils\builder.py", line 1021, in exec_ninja
    self.__execute(
  File "C:\gtk-build\github\gvsbuild\gvsbuild\utils\builder.py", line 1080, in __execute
    subprocess.check_call(args, cwd=working_dir, env=env, shell=True)
  File "C:\Python310\lib\subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'ninja install' returned non-zero exit status 1.
Error: lz4 build failed
```

**Pull request closed without merging**